### DISCUSSION TITLE
Are we generic yet?

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#6068] [Q&A]
Python 3.12 just came out and I was excited to play around with generics. When I try to declare a generic class:

```python
class Hello[T]:
    pass
```

I'm faced with:

```bash
Pyright: Class type parameter syntax requires Python 3.12 or newer
```

Is pyright 3.12 ready?

**Comment by erictraut** [comment]
Yes, pyright has supported Python 3.12 (and PEP 695 in particular) for almost a year already. Most likely, you don't have the correct Python interpreter selected. Or you have explicitly overridden the `pythonVersion` in your config file.

**Reply by ghost** [reply]
I'm absolutely sure that I have the right python version and I did not override the python version manually. I'm always using a virtual environment. I attached a screen shot with an open interpreter session:

<img width="2540" alt="image" src="https://github.com/microsoft/pyright/assets/75550932/2ed83863-0684-41c2-b5e3-218c2d884195">

**Reply by ghost** [reply]
Apparently, this is something with neovim not recognizing the actual python version. For future reference, I solved it temporarily by executing the following snippet in the LSP client's `on_attach`:

```lua
        local handle = io.popen "which python"
        if not handle then return end
        local output = handle:read("*a"):gsub("[\n\r]", "")
        handle:close()
        vim.defer_fn(function()
            vim.cmd("PyrightSetPythonPath " .. output)
        end, 1)
```

**Reply by wookayin** [reply]
@westernwontons Thanks for posting this. I think this is a neovim-specific issue (which is however very helpful information!), but I was also wondering why pyright was not picking up the right python in this situation without explicitly calling `workspace/didChangeConfiguration`.

(1) I have python3.12 foremost in the `$PATH` -- this is not only for the neovim but all subsequent subprocesses 
(2) I configured lspconfig and mason to use `pyright-languageserver` running with the same python3.12 interpreter
(3) using the default setting for lspconfig.pyright.setup
```
(1) $PATH == "$HOME/.mambaforge/envs/py312:..."

(2) :LspInfo
Client: pyright (id: 1, bufnr: [19])
 	filetypes:       python
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             $HOME/.mambaforge/envs/py312/bin/pyright-langserver --stdio

(3) :lua= vim.lsp.get_clients({ name = 'pyright' })[1].config.settings
{
  python = {
    analysis = {
      autoSearchPaths = true,
      diagnosticMode = "openFilesOnly",
      typeCheckingMode = "basic",
      useLibraryCodeForTypes = true
    },
    -- no pythonPath = ... here, just let pyright pick up the default python from ENV
  }
}
```

What's the mechanism of pyright to choose the python interpreter or version if `python.pythonPath` was not given in the settings? I was thinking that this should be [the default python under the current environment](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md#configuring-your-python-environment) -- what `which python` gives as per `$PATH`.
